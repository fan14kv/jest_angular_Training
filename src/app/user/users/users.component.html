<h2>User Form</h2>
    <form [formGroup]="userForm" (ngSubmit)="onSubmit()">
      <label>
        Name:
        <input formControlName="name" />
      </label>
      <div *ngIf="userForm.get('name')?.invalid && userForm.get('name')?.touched">
        Name is required.
      </div>

      <label>
        Email:
        <input formControlName="email" />
      </label>
      <div *ngIf="userForm.get('email')?.invalid && userForm.get('email')?.touched">
        Valid email is required.
      </div>

      <button appDisableOnClick type="" [disabled]="userForm.invalid">Submit</button>
    </form>

    <p *ngIf="submitted()">{{successMessage() | truncate:5}}</p>
