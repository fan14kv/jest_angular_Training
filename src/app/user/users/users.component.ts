import { CommonModule } from '@angular/common';
import { Component, signal } from '@angular/core';
import { FormControl, FormGroup, ReactiveFormsModule, Validators } from '@angular/forms';
import { DisableOnClickDirective } from '../../shared/directives/disable-on-click.directive';
import { TruncatePipe } from "../../shared/pipes/truncate.pipe";

@Component({
  selector: 'app-users',
    standalone: true,
    imports: [ReactiveFormsModule,
    CommonModule, DisableOnClickDirective, TruncatePipe],
  templateUrl: './users.component.html',
  styleUrls: ['./users.component.css']
})
export class UsersComponent {
  dndStatusForm!: FormGroup;
  submitted = signal<boolean>(false);
  successMessage = signal<string>('Form submitted successfully!');
  userForm = new FormGroup({
    name: new FormControl('', [Validators.required]),
    email: new FormControl('', [Validators.required, Validators.email])
  });

  onSubmit() {
    this.submitted.set(true);
    console.log(this.userForm.value);
  }

  ///Form ...
}
