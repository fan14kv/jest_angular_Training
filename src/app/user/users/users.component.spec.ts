import { ComponentFixture, TestBed } from "@angular/core/testing"
import { UsersComponent } from "./users.component"
import { Component } from "@angular/core";

describe("UsersComponent",()=>{
  let fixture : ComponentFixture<UsersComponent>;
  let component : UsersComponent;
  beforeEach( async()=>{
    await TestBed.configureTestingModule({
      imports:[UsersComponent ]
    }).compileComponents();

    fixture = TestBed.createComponent(UsersComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  describe('submit button',()=>{
    it("should disable the submit button when form is invalid",()=>{
      const button:HTMLButtonElement = fixture.nativeElement.querySelector('button');
      expect(button.disabled).toBe(true);
    });

    it("should Enable the submit button when form is valid",()=>{
      const button:HTMLButtonElement = fixture.nativeElement.querySelector('button');
      component.userForm.setValue({name:"Kumar",email:"kumar@gmail.com"});
      fixture.detectChanges();
      
      expect(button.disabled).toBe(false);
    });
  });
   describe('submit Form',()=>{
    it("should invalied when the field are empty",()=>{
      expect(component.userForm.valid).toBe(false);
      expect(component.userForm.get('name')?.errors?.['required']).toBe(true);
      expect(component.userForm.get('email')?.errors?.['required']).toBe(true);
    });

    it("should invalied when the field are empty",()=>{
      component.userForm.setValue({name:"Kumar",email:"kumar@gmail.com"});
      expect(component.userForm.valid).toBe(true);
      expect(component.userForm.get('name')?.errors?.['required']).toBeFalsy();
      expect(component.userForm.get('email')?.errors?.['required']).toBeFalsy();
    });
  });
   describe('OnSubmit',()=>{
    it('should set submit to true',()=>{
      const consoleSpy = jest.spyOn(console,'log');
      component.userForm.setValue({name:"Kumar",email:"kumar@gmail.com"});
      component.onSubmit();
      expect(component.submitted()).toBe(true);
      expect(consoleSpy).toHaveBeenCalledWith({name:"Kumar",email:"kumar@gmail.com"});

    });
   });

   it('should disable when the button is clicked',()=>{
    component.userForm.setValue({name:"Kumar",email:"kumar@gmail.com"});
    fixture.detectChanges();
    const button:HTMLButtonElement = fixture.nativeElement.querySelector('button');
    expect(button.disabled).toBe(false);
    button.click();
    fixture.detectChanges();
    expect(button.disabled).toBe(true);
   })
})