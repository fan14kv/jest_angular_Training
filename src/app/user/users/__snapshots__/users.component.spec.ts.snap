// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersComponent submit button should Enable the submit button when form is valid 1`] = `
<app-users
  submitted={[Function Function]}
  successMessage={[Function Function]}
  userForm={[Function FormGroup]}
>
  <h2>
    User Form
  </h2><form
    class="ng-untouched ng-pristine ng-valid"
    novalidate=""
  >
    <label>
       Name: 
      <input
        class="ng-untouched ng-pristine ng-valid"
        formcontrolname="name"
      />
    </label>
    <label>
       Email: 
      <input
        class="ng-untouched ng-pristine ng-valid"
        formcontrolname="email"
      />
    </label>
    <button
      appdisableonclick=""
      type="submit"
    >
      Submit
    </button>
  </form>
</app-users>
`;
